numpy/core/src/umath/loops_unary_fp.dispatch.c.src

```diff
@@ -0,0 +1,228 @@
+    const int sdst = 1;
+#else
+    const npy_intp sdst = steps[1] / sizeof(dst[0]);
+#endif
+#if @VCHK@
+    const int vstep = npyv_nlanes_@sfx@;
+    // unroll by 4 vectors
+    const int wstep = vstep * 4;
+    for (; len >= wstep; len -= wstep, src += ssrc*wstep, dst += sdst*wstep) {
+        /**begin repeat3
+         * #N  = 0, 1, 2, 3#
+         */
+        #if @STYPE@ == CONTIG
+            npyv_@sfx@ v_src@N@ = npyv_load_@sfx@(src + vstep*@N@);
+        #else
+            npyv_@sfx@ v_src@N@ = npyv_loadn_@sfx@(src + ssrc*vstep*@N@, ssrc);
+        #endif
+        /**end repeat3**/
+        /**begin repeat3
+         * #N  = 0, 1, 2, 3#
+         */
+        npyv_@sfx@ v_unary@N@ = npyv_@intr@_@sfx@(v_src@N@);
+        /**end repeat3**/
+        /**begin repeat3
+         * #N  = 0, 1, 2, 3#
+         */
+        #if @DTYPE@ == CONTIG
+            npyv_store_@sfx@(dst + vstep*@N@, v_unary@N@);
+        #else
+            npyv_storen_@sfx@(dst + sdst*vstep*@N@, sdst, v_unary@N@);
+        #endif
+        /**end repeat3**/
+    }
+    for (; len > 0; len -= vstep, src += ssrc*vstep, dst += sdst*vstep) {
+    #if @STYPE@ == CONTIG
+        #if @repl_0w1@
+            npyv_@sfx@ v_src0 = npyv_load_till_@sfx@(src, len, 1);
+        #else
+            npyv_@sfx@ v_src0 = npyv_load_tillz_@sfx@(src, len);
+        #endif
+    #else
+        #if @repl_0w1@
+            npyv_@sfx@ v_src0 = npyv_loadn_till_@sfx@(src, ssrc, len, 1);
+        #else
+            npyv_@sfx@ v_src0 = npyv_loadn_tillz_@sfx@(src, ssrc, len);
+        #endif
+    #endif
+        npyv_@sfx@ v_unary0 = npyv_@intr@_@sfx@(v_src0);
+    #if @DTYPE@ == CONTIG
+        npyv_store_till_@sfx@(dst, len, v_unary0);
+    #else
+        npyv_storen_till_@sfx@(dst, sdst, len, v_unary0);
+    #endif
+    }
+#else
+    // unroll by 4 scalars for non-supported archs
+    #ifndef NPY_DISABLE_OPTIMIZATION
+    for (; len >= 4; len -= 4, src += ssrc*4, dst += sdst*4) {
+        /**begin repeat3
+         * #N  = 0, 1, 2, 3#
+         */
+        const npyv_lanetype_@sfx@ src@N@ = src[ssrc*@N@];
+        /**end repeat3**/
+        /**begin repeat3
+         * #N  = 0, 1, 2, 3#
+         */
+        dst[sdst*@N@] = c_@intr@_@sfx@(src@N@);
+        /**end repeat3**/
+    }
+    #endif // NPY_DISABLE_OPTIMIZATION
+    // collect breadcrumbs
+    for (; len > 0; --len, src += ssrc, dst += sdst) {
+        const npyv_lanetype_@sfx@ src0 = *src;
+        *dst = c_@intr@_@sfx@(src0);
+    }
+#endif // @VCHK@
+}
+/**end repeat2**/
+/**end repeat1**/
+/**end repeat**/
+
+/********************************************************************************
+ ** Defining ufunc inner functions
+ ********************************************************************************/
+/**begin repeat
+ * #TYPE = FLOAT, DOUBLE#
+ * #sfx  = f32, f64#
+ */
+/**begin repeat1
+ * #kind  = sqrt, absolute, square, reciprocal#
+ * #intr  = sqrt, abs,      square, recip#
+ * #clear = 0,    1,        0,      0#
+ */
```

Comment:
Remind me what the macro `NPY_CPUT_DISPATCH_CURFX` does ? The only reference in the docs is [here] (https://numpy.org/devdocs/reference/simd/simd-optimizations.html?highlight=npy_cpu_dispatch_curfx). I thought we do all the necessary dispatching in `numpy/core/code_generators/generate_umath.py`, here we should only create the baseline and dispatch versions of the loops.

Reply:
The dispatching mechanism is divided into three steps:
 1. forward declarations for the dispatched functions, that can be done via `NPY_CPU_DISPATCH_DECLARE`
 2. defining the dispatched functions, that require a dispatch-able source and `NPY_CPU_DISPATCH_CURFX`
 3. runtime dispatching call, according to the highest available feature, that can be done via `NPY_CPU_DISPATCH_CALL`
   or `NPY_CPU_DISPATCH_CALL_XB`

all the above-mentiond macros are defined and explained in [npy_cpu_dispatch.h](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/npy_cpu_dispatch.h).

> Remind me what the macro NPY_CPU_DISPATCH_CURFX does ?

**CURFX** stands for "Current suffix", and this macro is used to appends the current suffix of the target
feature to a ceratin token, since dispatch-able sources are compiled multiple times according to
the required targets which lead to duplicate the exported symbols if there's no such macro
like `NPY_CPU_DISPATCH_CURFX`.

> we do all the necessary dispatching in numpy/core/code_generators/generate_umath.py, here we should only create the baseline and dispatch versions of the loops.

`numpy/core/code_generators/generate_umath.py` only contains the third step (runtime dispatching call), which happens during the initializing of umath module.